### DISCUSSION TITLE
Test Cases Created From OpenAPI spec not adhering to specifications

### DISCUSSION THREAD

**Discussion by AnirudhChaudhary** [Cornutum/tcases#200] [Q&A] [answered]
Hello Kerry,

I am trying to create test cases from my OpenAPI spec but am running into troubles because I am getting test cases that do not adhere to the openapi specifications. 

This is happening particularly for a variable that is coming as null when I have set nullable to false. 

This is happening in a test called getAPI_XisDefined_Is_Yes(). I have clearly stated that X should not be nulled out in the openapi spec under the "schema" section. However, in the test I stated above, X is nulled out. I am curious about what is happening in this test and if its a mistake in the specification document or something that TCases is not recognizing.

Thanks.

**Comment by kerrykimbrough** [comment]
Can you attach the OpenAPI file so I can reproduce this?

Also, have you looked at messages in the log file? This is usually a file named tcases-api-test.log.

**Reply by kerrykimbrough** [reply]
If X is type=string and minLength is undefined or 0, then an empty string is valid.

**Reply by AnirudhChaudhary** [reply]
This is a screenshot of the section of the openapi spec that is causing the problem.
![StudentID_Nullable_OpenAPI](https://user-images.githubusercontent.com/52807055/126390182-232dcb12-6bf0-416b-9674-80e3c0524ff5.JPG)

**Reply by kerrykimbrough** [reply]
Please open a new issue. Be sure to attach a copy of the OpenAPI file and an explanation of why the results you are seeing are wrong.

**Reply by kerrykimbrough** [reply]
Also attach a copy of the tcases-api-test.log file and include the name of the OpenAPI file.

**Reply by AnirudhChaudhary** [reply]
Hi Kerry, I am currently running this project with the maven plugin. Usually when I ran TCases, I would see the cases-api-test.log file but while running it through the maven plugin, I do not see that file. Does it have a different name?

Also I apologize, but I do not currently have the ability to send the entire OpenAPI file, simply only a screenshot. Will this significantly inhibit your ability to diagnose and solve the problem?

**Reply by kerrykimbrough** [reply]
Let me guess: you are seeing a test case for `GET /api/StudentPerf?StudentIDS=` that expects an `isSuccess()` response. But by default `nullable` is false for this parameter. So there is another test case that sends the same request but expects an `isBadRequest()` response. 

Is that the problem you are seeing?

**Reply by AnirudhChaudhary** [reply]
![StudentID_test_case](https://user-images.githubusercontent.com/52807055/126539881-fde152a9-7e36-4bd0-897a-5a36942b04ab.jpg)
The StudentID variable in this test case is nulled out but it should not be because it is a required parameter for the request. My problem: Why is this test getting created when the parameter should not be nulled out from the spec? And if it is getting created, why is it expecting a Success?

**Comment by kerrykimbrough** [comment, marked as answer]
This problem happens when TFOA generates inconsistent test cases for null vs. empty values -- see issue #201.

In the case shown above, you can work around the problem by adding `"minItems": 1` to the schema for StudentIDs. This tells TFOA that an empty array (which in this case looks that same as a null array) is not a valid value. This is not a perfect solution, because it will also result in redundant failure test cases -- one for an invalid null value and another for an invalid empty value.

**Reply by kerrykimbrough** [reply]
Now fixed in release 3.7.1 -- see the [release notes](https://github.com/Cornutum/tcases/tree/master/ReleaseNotes.md#371) for details.

**Reply by kerrykimbrough** [reply]
BTW, it turns out that the correct encoding for an empty StudentIDs array actually is `StudentIDs=`, i.e. an empty value. For a null value, the encoding would be different: just `StudentIDs` (no `=`). So even though issue #201 was a real problem that is now fixed, you will see the same result in 3.7.1.
